### PR TITLE
Fixes Negative numbers for @ not being recognized

### LINKED ISSUE #40: Negative numbers for @ are not being recognized
While "@0" and "@positive" integer work, "@-1" does not work. 

On GNU date it returns the time in seconds before JAN 01 1970.

For example:

```
"date -d '@-1'"     
Thu Jan  1 08:59:59 JST 1970
```

```
"date -d '@-1000000'"           
Sat Dec 20 19:13:20 JST 1969
```

### PR DESCRIPTION
**Pull request by philolo1**
This commit resolves #40.
Adds new file and functions parse_timestamp.
Adds tests for handling negative numbers.

@tertsdiepraam @sylvestre

### CODE REVIEW

File: src/parse_timestamp.rs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    None
+}
+
+#[cfg(test)]
+mod tests {
+
+    use crate::parse_timestamp::parse_timestamp;
+
+    #[test]
+    fn test_valid_timestamp() {
```
Comment (tertsdiepraam) [inline src/parse_timestamp.rs:36]: Nice tests!

**Review by tertsdiepraam — COMMENTED** [review]
Lovely! I have some minor comments, but it's looking pretty good!

File: src/parse_timestamp.rs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    if let Ok((_, (_, sign, number_str))) = res {
+        let number = format!("{}{}", sign, number_str).parse();
+
+        return match number {
+            Err(_) => None,
+            Ok(val) => Some(val),
+        };
+    }
+
+    None
```
[the remark this answers is not in the document]
Comment (philolo1) [reply src/parse_timestamp.rs:27]: That will make code easier, thanks.

**Comment by philolo1** [comment]
@tertsdiepraam adjusted the code.

File: src/parse_timestamp.rs [hunk 3]
```diff
@@ -0,0 +1,67 @@
+use nom::branch::alt;
+use nom::character::complete::{char, digit1};
+use nom::combinator::all_consuming;
+use nom::multi::fold_many0;
+use nom::sequence::preceded;
+use nom::sequence::tuple;
+use nom::{self, IResult};
+
+pub(crate) fn parse_timestamp(s: &str) -> Option<i64> {
```
Comment (sylvestre) [inline src/parse_timestamp.rs:9]: can we return:
Result<i64, ParseTimestampError>

like we return ParseDateTimeError elsewhere?

File: src/parse_timestamp.rs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    let s = s.trim().to_lowercase();
+    let s = s.as_str();
+
+    let res: IResult<&str, (char, &str)> = all_consuming(preceded(
+        char('@'),
+        tuple((
+            // Note: gnu date allows multiple + and - and only considers the last one
```
Comment (sylvestre) [inline src/parse_timestamp.rs:16]: maybe add that we are doing the same

File: src/parse_timestamp.rs [hunk 5]
```diff
@@ -0,0 +1,67 @@
+use nom::branch::alt;
```
Comment (sylvestre) [inline src/parse_timestamp.rs:1]: please add the license header

File: src/lib.rs [hunk 6]
```diff
@@ -169,11 +171,11 @@ pub fn parse_datetime_at_date<S: AsRef<str> + Clone>(
     }
 
     // Parse epoch seconds
-    if s.as_ref().bytes().next() == Some(b'@') {
-        if let Ok(parsed) = NaiveDateTime::parse_from_str(&s.as_ref()[1..], "%s") {
-            if let Ok(dt) = naive_dt_to_fixed_offset(date, parsed) {
-                return Ok(dt);
-            }
+    if let Some(timestamp) = parse_timestamp(s.as_ref()) {
+        let timestamp_date = NaiveDateTime::from_timestamp_opt(timestamp, 0).unwrap();
```
Comment (sylvestre) [inline src/lib.rs:175]: what happens in case of error ?

Comment (philolo1) [reply src/lib.rs]: Good catch!

**Comment by philolo1** [comment]
@sylvestre updated, let me know if you like to resolve the comments, or i should. I am not sure how you handle it.

File: src/parse_timestamp.rs (hunk 5 above)
Comment (philolo1) [reply src/parse_timestamp.rs:8]: Done

File: src/parse_timestamp.rs (hunk 4 above)
Comment (philolo1) [reply src/parse_timestamp.rs]: Done

File: src/parse_timestamp.rs (hunk 3 above)
Comment (philolo1) [reply src/parse_timestamp.rs]: Done

**Comment by philolo1** [comment]
@tertsdiepraam @sylvestre just checking on the status. I think someone needs to review it.

File: src/parse_timestamp.rs [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+    #[test]
+    fn test_valid_timestamp() {
+        assert_eq!(parse_timestamp("@1234"), Ok(1234));
+        assert_eq!(parse_timestamp("@99999"), Ok(99999));
+        assert_eq!(parse_timestamp("@-4"), Ok(-4));
+        assert_eq!(parse_timestamp("@-99999"), Ok(-99999));
+        assert_eq!(parse_timestamp("@+4"), Ok(4));
+        assert_eq!(parse_timestamp("@0"), Ok(0));
+
+        // gnu date acceppts numbers signs and uses the last sign
```
Comment (sylvestre) [inline src/parse_timestamp.rs:98]:
```suggestion
        // gnu date accepts numbers signs and uses the last sign
```

**Pull request merged**